docs/discussion/Cosign Upgrade Discussion 2024.md

```diff
@@ -0,0 +1,662 @@
+
+Compared to the `CertificateStore`, the `KeyManagementProvider` config spec could be updated to be more flexible. A new `name`` field will be used only in CLI scenarios to mirror CRD name functionality as a unique identifier. This enables multiple KMS of same type to be used.
+
+```yaml
+apiVersion: config.ratify.deislabs.io/v1beta1
+kind: KeyManagementProvider
+metadata:
+  name: ratify-notation-inline-cert-kmprovider
+  annotations:
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-weight: "5"
+spec:
+  type: inline
+  parameters:
+    contentType: key
+    value: |
+      ---------- BEGIN RSA KEY ------------
+      ******
+      ---------- END RSA KEY ------------
+```
+
+CLI Config
+```json
+{
+  ...
+  "keyManagementSystems": {
+    {
+      "name": "ratify-notation-inline-cert-kmprovider",
+      "type": "inline",
+      "contentType": "key",
+      "value": "---------- BEGIN RSA KEY ------------
+      ******
+      ---------- END RSA KEY ------------"
+    },
+    {
+      "name": "ratify-notation-inline-cert-kmprovider-2",
+      "type": "inline",
+      "contentType": "key",
+      "value": "---------- BEGIN RSA KEY ------------
+      ******
+      ---------- END RSA KEY ------------"
+    },
+  }
+  ...
+}
+```
+
+**Do we want to consider making the inline provider accept an array of content?**
+
+### Option 3: Introduce new CRD `KeyStore` separate from `CertificateStore`
+
+All of the same considerations as Option 1. It will be a completely new CRD with an inline and AKV provider equivalent.
+
+Pros: 
+- More intuitive naming: embedding 'key' specification in cert store might be confusing
+- No breaking changes or backwards compatability requirements
+Cons:
+- Yet one more resource type user must be concerned to configure
+- Providers will share almost the exact same implementation as certificate stores
+
+### How do we support CLI scenario?
+  - CLI mode should support specification of a list of keys within the cosign verifier. Cosign plugin will handle directory configuration and key reading within implementation. Currently, one a single `key` path can be provided. This will be updated to support an absolute `filePath` to a key per entry in the `keys` field of the Cosign trust policy (see trust policy details [here](#trust-policy)). For backward compatability, the existing `key` field will also be honored.
+    - example `config.json`:
+      ```json
+      {
+        ...
+        "verifier": {
+          "version": "1.0.0",
+          "plugins": [
+              {
+                  "type": "cosign",
+                  "name": "cosign-wabbit-networks",
+                  "artifactTypes": "application/vnd.dev.cosign.artifact.sig.v1+json",
+                  "trustPolicies": [
+                    {
+                      "name": "wabbit-networks-images",
+                      "keys": [
+                        {
+                          "provider": "wabbit-key-1", // this can be any name
```

Comment:
this property `provider` may confuse people as it may indicate KMS provider. Should we name it `certName`?